### PR TITLE
Misbehaviour Monitoring

### LINKED ISSUE #112: Misbehaviour monitoring & relay
Given a particular configuration option, using the local light clients, the relayer should periodically check if any conflicting (equivocating) headers have been submitted to any on-chain IBC light clients on the chains which are being monitored (relayed between), and if they have, the relayer should automatically submit proof-of-misbehaviour to freeze the on-chain light clients (and print a very scary red warning message encouraging the user to immediately investigate).

### PR DESCRIPTION
**Pull request by akhilkumarpilli**
closes: #112

### CODE REVIEW

File: relayer/naive-strategy.go [hunk 1]
```diff
@@ -232,6 +238,36 @@ func (nrs *NaiveStrategy) UnrelayedAcknowledgements(src, dst *Chain) (*RelaySequ
 
 // HandleEvents defines how the relayer will handle block and transaction events as they are emitted
 func (nrs *NaiveStrategy) HandleEvents(src, dst *Chain, events map[string][]string) {
+	// check for misbehaviour
```
Comment (colin-axner) [inline relayer/naive-strategy.go]: let's make this a helper function

File: relayer/naive-strategy.go [hunk 2]
```diff
@@ -232,6 +238,36 @@ func (nrs *NaiveStrategy) UnrelayedAcknowledgements(src, dst *Chain) (*RelaySequ
...
+	if hdrs, ok := events[fmt.Sprintf("%s.%s", updateCliTag, headerTag)]; ok {
+		for i, hdr := range hdrs {
+			clientIDs := events[fmt.Sprintf("%s.%s", updateCliTag, clientIDTag)]
+
+			if src.PathEnd.ClientID == clientIDs[i] && dst.PathEnd.ClientID == clientIDs[i] {
```
Comment (colin-axner) [inline relayer/naive-strategy.go]:
```suggestion
			if src.PathEnd.ClientID == clientIDs[i] {
```
I assume we are handling events emitted by source correct? If so then we only need to monitor misbehaviour for the client on source

File: relayer/naive-strategy.go [hunk 3]
```diff
@@ -232,6 +238,36 @@ func (nrs *NaiveStrategy) UnrelayedAcknowledgements(src, dst *Chain) (*RelaySequ
...
+				hdrBytes, err := hex.DecodeString(hdr)
+				if err == nil {
```
Comment (colin-axner) [inline relayer/naive-strategy.go]: Once this is a helper function, I'd use `err != nil` and return the err and then log it once in `HandleEvents`

File: relayer/naive-strategy.go [hunk 4]
```diff
@@ -232,6 +238,36 @@ func (nrs *NaiveStrategy) UnrelayedAcknowledgements(src, dst *Chain) (*RelaySequ
...
+					exportedHeader, err := clienttypes.UnmarshalHeader(src.Encoding.Marshaler, hdrBytes)
+					if err == nil {
+						gotHeader, ok := exportedHeader.(*tmclient.Header)
```
Comment (colin-axner) [inline relayer/naive-strategy.go]:
```suggestion
						emittedHeader, ok := exportedHeader.(*tmclient.Header)
```

I like this naming slightly better

File: relayer/naive-strategy.go [hunk 5]
```diff
@@ -232,6 +238,36 @@ func (nrs *NaiveStrategy) UnrelayedAcknowledgements(src, dst *Chain) (*RelaySequ
...
+						if ok {
+							trustedHeader, err := src.GetLatestLightHeader()
```
Comment (colin-axner) [inline relayer/naive-strategy.go]: We need to compare [consensus states](https://github.com/cosmos/ibc-go/blob/333c1f338b2a14a1928a6f8ab64c37123c0e97b6/modules/light-clients/07-tendermint/types/header.go#L18). We should check the consensus state of the emitted header against a header which we construct/retrieve from our light client.

You'll likely need to use this [function](https://github.com/cosmos/relayer/blob/master/relayer/ibc-client.go#L112) and passing in the height of the emitted header. Then use the reflect DeepEqual check against the consensus states

File: relayer/naive-strategy.go [hunk 6]
```diff
@@ -232,6 +238,36 @@ func (nrs *NaiveStrategy) UnrelayedAcknowledgements(src, dst *Chain) (*RelaySequ
...
+							if err == nil && !reflect.DeepEqual(gotHeader, trustedHeader) {
+								misbehaviour := tmclient.NewMisbehaviour(clientIDs[i], gotHeader, trustedHeader)
```
Comment (colin-axner) [inline relayer/naive-strategy.go]: lets make `clientIDs[i]` a var at the beginning of the for, maybe a name like `emittedClientID`

Comment (colin-axner) [inline relayer/naive-strategy.go]: before we pass in the headers. At this point the emittedHeader is good to go, but the header we created is missing trusted header info, this is where it gets tricky. If the relayer misses some blocks, it is possible it misses some previous misbehaving header, meaning the trusted header used in the emittedHeader might be invalid. 

We should create another helper function which gets all of the consensus states for this client, iterates through them and checks to see if they are valid (you should probably deduplicate some of the logic here). Once you find the earliest instance of misbehaviour. You need to reconstruct the misbehaving header via tx events and use the trusted header from the consensus state which occurs directly before that

This is kinda complicated, so lets leave it off for now and open an issue for it when this is ready to be merged

File: relayer/naive-strategy.go [hunk 7]
```diff
@@ -232,6 +238,36 @@ func (nrs *NaiveStrategy) UnrelayedAcknowledgements(src, dst *Chain) (*RelaySequ
...
+						}
+					}
+
+				} else {
+					src.Error(err)
+				}
+			}
+		}
+	}
+
```
Comment (colin-axner) [inline relayer/naive-strategy.go:240]: We should try to write integration tests for this (we will need to update docker images)

Here is a description of how it should work:

Create two chains with the same validator and same chain-ids (we might need to refactor integration test code). Create a third chain with different validator/chain-id, construct links from chainA1 -> chainB and chainA2 -> B  (client, connection, channel). This should cause a divergence in state. 

Change one of the chainA's config to begin using the opposing client and submits an update message using its local chain state. The validator is still the same so verification should pass update checks. The chain who didn't change should detect misbehaviour and freeze the client. If successful, both chains switch to the still active client (on chainA) and the other chain attempts to cause misbehaviour (redo steps above)

Both clients on chainA's should become frozen if both implementations work

Let me know if you have questions

**Review by colin-axner — COMMENTED** [review]
Nice work

File: relayer/naive-strategy.go [hunk 8]
```diff
@@ -232,12 +237,68 @@ func (nrs *NaiveStrategy) UnrelayedAcknowledgements(src, dst *Chain) (*RelaySequ
...
+		src.Error(err)
+	}
+
 	rlyPackets, err := relayPacketsFromEventListener(src.PathEnd, dst.PathEnd, events)
 	if len(rlyPackets) > 0 && err == nil {
 		nrs.sendTxFromEventPackets(src, dst, rlyPackets)
 	}
 }
 
+func checkAndSubmitMisbehaviour(src *Chain, events map[string][]string) error {
```
Comment (colin-axner) [inline relayer/naive-strategy.go]: lets add some godoc, this is a hefty function

File: relayer/naive-strategy.go [hunk 9]
```diff
@@ -232,12 +237,68 @@ func (nrs *NaiveStrategy) UnrelayedAcknowledgements(src, dst *Chain) (*RelaySequ
...
+				if err != nil {
+					return err
+				}
+
+				emittedHeader, ok := exportedHeader.(*tmclient.Header)
+				if !ok {
+					return fmt.Errorf("emitted header is not tendermint type")
+				}
+
+				trustedHeader, err := src.GetLightSignedHeaderAtHeight(emittedHeader.Header.Height)
```
Comment (colin-axner) [inline relayer/naive-strategy.go]: So, this function currently returns a header without a trusted validator set and trusted height. We need to inject these into the header. 

This gets a little complicated. It is possible the trusted header used to do a valid update of `emittedHeader` is also misbehavior. We need to iterate through all consensus states and check that all the previous submitted updates are valid. If this occurs we need to parse the tx that created the past misbehaviour and construct misbehaviour against that.

Lets handle this case in a different pr. For now, take the trusted validator set and height of the emitted header and set those value for the trusted header

```go
trustedHeader.TrustedValidators = emittedHeader.TrustedValidators
trustedHeader.TrustedHeight = emittedHeader.TrustedHeight
```

**Review by colin-axner — COMMENTED** [review]
Woohoo! Looks great! Still a lot of more complexity we need to add but the building blocks are there. Lets get integration tests in and then merge this and go from there

File: relayer/naive-strategy.go (hunk 8 above)
Comment (colin-axner) [inline relayer/naive-strategy.go]: also lets put this in its own file `misbehaviour.go`

File: relayer/misbehaviour.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+var (
+	// strings for parsing events
+	updateCliTag = "update_client"
+	headerTag    = "header"
+	clientIDTag  = "client_id"
+)
+
+// checkAndSubmitMisbehaviour check headers from update_client tx events
+// and submit misbehaviour if found any
```
Comment (colin-axner) [inline relayer/misbehaviour.go]:
```suggestion
// checkAndSubmitMisbehaviour check headers from update_client tx events
// against the associated light client. If the headers do not match, the emitted 
// header and a reconstructed header are used in misbehaviour submission to
// the IBC client on the source chain. 
```

File: relayer/misbehaviour.go [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+					return fmt.Errorf("emitted header is not tendermint type")
+				}
+
+				trustedHeader, err := src.GetLightSignedHeaderAtHeight(emittedHeader.Header.Height)
+				if err != nil {
+					return err
+				}
+
+				trustedHeader.TrustedValidators = emittedHeader.TrustedValidators
+				trustedHeader.TrustedHeight = emittedHeader.TrustedHeight
```
Comment (colin-axner) [inline relayer/misbehaviour.go]: these should be the first lines inside the if statement

File: relayer/misbehaviour.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	headerTag    = "header"
+	clientIDTag  = "client_id"
+)
+
+// checkAndSubmitMisbehaviour check headers from update_client tx events
+// against the associated light client. If the headers do not match, the emitted
+// header and a reconstructed header are used in misbehaviour submission to
+// the IBC client on the source chain.
+func checkAndSubmitMisbehaviour(src *Chain, events map[string][]string) error {
+	if hdrs, ok := events[fmt.Sprintf("%s.%s", updateCliTag, headerTag)]; ok {
```
Comment (fedekunze) [inline relayer/misbehaviour.go]: I think this is more readable if you use `!ok` and return `nil` instead

File: relayer/misbehaviour.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+		for i, hdr := range hdrs {
+			clientIDs := events[fmt.Sprintf("%s.%s", updateCliTag, clientIDTag)]
+			emittedClientID := clientIDs[i]
```
Comment (fedekunze) [inline relayer/misbehaviour.go]: this will panic if there string array is empty

File: relayer/misbehaviour.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+			if src.PathEnd.ClientID == emittedClientID {
```
Comment (fedekunze) [inline relayer/misbehaviour.go]: ditto use `!=` and `continue` for better readibility

File: relayer/misbehaviour.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+				hdrBytes, err := hex.DecodeString(hdr)
+				if err != nil {
+					return err
```
Comment (fedekunze) [inline relayer/misbehaviour.go]: i think we should wrap the errors with a note and the client id

File: relayer/misbehaviour.go [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+				}
+
+				exportedHeader, err := clienttypes.UnmarshalHeader(src.Encoding.Marshaler, hdrBytes)
+				if err != nil {
+					return err
```
Comment (fedekunze) [inline relayer/misbehaviour.go]: ditto

File: relayer/misbehaviour.go [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+				if !ok {
+					return fmt.Errorf("emitted header is not tendermint type")
+				}
+
+				trustedHeader, err := src.GetLightSignedHeaderAtHeight(emittedHeader.Header.Height)
+				if err != nil {
+					return err
+				}
+
+				if !IsMatchingConsensusState(emittedHeader.ConsensusState(), trustedHeader.ConsensusState()) {
```
Comment (fedekunze) [inline relayer/misbehaviour.go]: ditto `continue`

File: relayer/misbehaviour.go [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+					trustedHeader.TrustedValidators = emittedHeader.TrustedValidators
+					trustedHeader.TrustedHeight = emittedHeader.TrustedHeight
+
+					misbehaviour := tmclient.NewMisbehaviour(emittedClientID, emittedHeader, trustedHeader)
+					msg, err := clienttypes.NewMsgSubmitMisbehaviour(emittedClientID, misbehaviour, src.MustGetAddress())
+					if err != nil {
+						return err
+					}
+					res, success, err := src.SendMsg(msg)
```
Comment (fedekunze) [inline relayer/misbehaviour.go]: validate msg before sending?

**Review by fedekunze — APPROVED** [review]

**Review by cwgoes — APPROVED** [review]
ACK, looks solid to me

Note that this is not incentive-compatible (also, the relayer may waste lots of gas relaying nonsense headers), but I think this is fine for now.

**Review by colin-axner — APPROVED** [review]
ACK, I will handle integration tests in a followup

**Pull request merged**